onnxruntime/core/graph/constant_folding.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/graph/constant_folding.h"
+#include "core/graph/graph_utils.h"
+#include "core/session/inference_session.h"
+#include "core/graph/model.h"
+#include "core/framework/ml_value.h"
+
+namespace onnxruntime {
+
+Status ConstantFolding::Apply(Graph& graph, Node& node, bool& modified) {
+  // TODO Can we reuse the model across calls of the constant folding rule?
+  // TODO Instead of computing one node at a time, we can traverse the whole graph and
+  // find constant parts of it. Then we can create bigger subgraphs to compute directly.
+  // This will be a useful optimization if we expect big subgraphs to be constant, and
+  // should be done through a Transformer, not a RewriteRule.
+  auto p_model = std::make_unique<onnxruntime::Model>("ConstantFoldingModel", false, ModelMetaData(),
+                                                      IOnnxRuntimeOpSchemaRegistryList({graph.GetSchemaRegistry()}),
+                                                      graph.DomainToVersionMap());
+  Graph& subgraph = p_model->MainGraph();
+
+  std::vector<onnxruntime::NodeIndex> subgraph_nodes;
+  subgraph_nodes.push_back(node.Index());
+
+  // Build the subgraph.
+  graph_edit_utils::BuildSubgraph(graph, subgraph_nodes, subgraph);
+
+  SessionOptions so;
+  so.session_logid = "ConstantFoldingSession";
+  // Disable default graph transformers for the constant node (can lead to infinite recursion otherwise).
+  so.enable_default_transformers = false;
+  InferenceSession session_object{so};
+  // TODO Make it possible to pass the model directly to the session instead of having to dump it to a stream.
+  std::stringstream model_istream;
+  p_model->ToProto().SerializeToOstream(&model_istream);
+  ONNXRUNTIME_RETURN_IF_ERROR(session_object.Load(model_istream));
+  ONNXRUNTIME_RETURN_IF_ERROR(session_object.Initialize());
+
+  // Execute the subgraph. No inputs are needed as they are all initializers.
+  std::vector<std::string> output_names;
+  for (auto& output : subgraph.GetOutputs()) {
+    output_names.push_back(output->Name());
+  }
+  std::unordered_map<std::string, MLValue> feeds;
+  std::vector<MLValue> fetches;
+
+  Status st = session_object.Run(feeds, output_names, &fetches);
```

Comment:
What if some node is not deterministic?

Reply:
That's a good point, Changming. In that case, we should disable constant folding I believe. Is there an easy way to find out which operators are non-deterministic?

Reply:
@linkerzhang Any suggestion?

Reply:
Not sure about a systematic way to find all. I think the "generator" ops (all the Random* and Multinomial) are primarily the non-deterministic ones.

Reply:
Is running it twice and checking if the outputs are the same a brute force but usable approach?

---
In reply to: [242302453](https://github.com/Microsoft/onnxruntime/pull/168#discussion_r242302453) [](ancestors = 242302453)

Reply:
I can do the running it twice, but I am worried that we will be adding too much overhead on large graphs. I can add some op types that are excluded from constant folding in the satisfyCondition() instead.
What do you think?